### PR TITLE
Fulltext filter sort

### PR DESCRIPTION
**Pull request by flametuner**
I explained more my usecase why I need this feature on #4412.

There's also another issue #1823, where this PR can be useful.

I started working on the codebase in the last 2 days so I may not see all the possibles edge cases where this code can fail, feel free to review.

### CODE REVIEW

**Comment by flametuner** [comment]
The filtering part was almost trivial, but the sorting part was really trick I hope this code can be useful.

**Review by lutter — CHANGES_REQUESTED** [review]
I've only had time to review the first commit - thanks for that, not being able to filter by more than just text has been a longstanding issue.

There should be a test for that functionality; the best place to put that is probably [here](https://github.com/graphprotocol/graph-node/blob/master/graphql/tests/query.rs) since that will test the full code path of running a GraphQL query with this new feature.

For the sorting part, I'll need some more time to review that; from what I have seen, the main point is to make it possible to order by more than one column. That has some implications for the cursor work ([draft](https://github.com/graphprotocol/graph-node/pull/3214))

I think it would be best to split this PR into two: one that fixes filtering, and one that adds multi-column sorting

File: graphql/src/store/query.rs [hunk 1]
```diff
@@ -174,18 +174,33 @@ fn build_filter(
...
+        entity_filter.push(filter);
+    }
+    if let Some(filter) = where_filter? {
+        entity_filter.extend(filter);
+    }
+
+    match entity_filter.len() {
+        0 => Ok(None),
+        1 => Ok(entity_filter.pop()),
+        _ => Ok(Some(EntityFilter::And(entity_filter)))
```
[the remark this answers is not in the document]
Comment (flametuner) [reply graphql/src/store/query.rs]: The `where_filter` is currently returnning a `Vec<EntityFilter>`, but I think I can chain `EntityFilter::And([text_filter, EntityFilter::And(where_filter)])` with no problems.

File: graphql/src/store/query.rs [hunk 2]
```diff
@@ -174,18 +174,33 @@ fn build_filter(
...
+        Some(r::Value::Null) | None => Ok(None),
+        _ => Err(QueryExecutionError::InvalidFilterError),
+    };
+
+    
+    let text_filter = match field.argument_value("text") {
+        Some(r::Value::Object(filter)) => build_fulltext_filter_from_object(filter),
+        None => Ok(None),
         _ => Err(QueryExecutionError::InvalidFilterError),
+    };
```
[the remark this answers is not in the document]
Comment (flametuner) [reply graphql/src/store/query.rs]: Done.

File: graphql/src/store/query.rs [hunk 3]
```diff
@@ -174,18 +174,33 @@ fn build_filter(
...
         },
-        Some(r::Value::Null) => Ok(None),
-        None => match field.argument_value("text") {
-            Some(r::Value::Object(filter)) => build_fulltext_filter_from_object(filter),
-            None => Ok(None),
-            _ => Err(QueryExecutionError::InvalidFilterError),
-        },
+        Some(r::Value::Null) | None => Ok(None),
+        _ => Err(QueryExecutionError::InvalidFilterError),
+    };
```
[the remark this answers is not in the document]
Comment (flametuner) [reply graphql/src/store/query.rs]: Done.

**Comment by flametuner** [comment]
> I think it would be best to split this PR into two: one that fixes filtering, and one that adds multi-column sorting

@lutter, I'll close this PR and create another one for just filter and the test. After the filter (and cursor) is merged, I can create another with the sorting.

**Pull request closed without merging**